x/ibc/04-channel/keeper/grpc_query_test.go

```diff
@@ -0,0 +1,655 @@
+		{
+			"empty request",
+			func() {
+				req = nil
+			},
+			false,
+		},
+		{
+			"invalid ID",
+			func() {
+				req = &types.QueryPacketCommitmentsRequest{
+					PortID:    "",
+					ChannelID: "test-channel-id",
+				}
+			},
+			false,
+		},
+		{
+			"success, empty res",
+			func() {
+				expCommitments = []*types.PacketAckCommitment{}
+
+				req = &types.QueryPacketCommitmentsRequest{
+					PortID:    "test-port-id",
+					ChannelID: "test-channel-id",
+					Req: &query.PageRequest{
+						Key:        nil,
+						Limit:      2,
+						CountTotal: true,
+					},
+				}
+			},
+			true,
+		},
+		{
+			"success",
+			func() {
+				_, _, _, _, channelA, _ := suite.coordinator.Setup(suite.chainA, suite.chainB)
+				commitment := types.NewPacketAckCommitment(channelA.PortID, channelA.ID, 1, []byte("hash"))
+				suite.chainA.App.IBCKeeper.ChannelKeeper.SetPacketCommitment(suite.chainA.GetContext(), channelA.PortID, channelA.ID, 1, commitment.Hash)
+
+				expCommitments = []*types.PacketAckCommitment{&commitment}
+				req = &types.QueryPacketCommitmentsRequest{
+					PortID:    channelA.PortID,
+					ChannelID: channelA.ID,
+					Req: &query.PageRequest{
+						Key:        nil,
+						Limit:      1,
+						CountTotal: false,
+					},
+				}
+			},
+			true,
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(fmt.Sprintf("Case %s", tc.msg), func() {
+			suite.SetupTest() // reset
+
+			tc.malleate()
+			ctx := sdk.WrapSDKContext(suite.chainA.GetContext())
+
+			res, err := suite.chainA.QueryServer.PacketCommitments(ctx, req)
+
+			if tc.expPass {
+				suite.Require().NoError(err)
+				suite.Require().NotNil(res)
+				suite.Require().Equal(expCommitments, res.Commitments)
+			} else {
+				suite.Require().Error(err)
+			}
+		})
+	}
+}
+
+func (suite *KeeperTestSuite) TestQueryUnrelayedPackets() {
+	var (
+		req    *types.QueryUnrelayedPacketsRequest
+		expSeq = []uint64{}
+	)
+
+	testCases := []struct {
+		msg      string
+		malleate func()
+		expPass  bool
+	}{
+		{
+			"empty request",
+			func() {
+				req = nil
+			},
+			false,
+		},
+		{
+			"invalid port ID",
+			func() {
+				req = &types.QueryUnrelayedPacketsRequest{
+					PortID:    "",
+					ChannelID: "test-channel-id",
+				}
+			},
+			false,
+		},
+		{
+			"invalid channel ID",
+			func() {
+				req = &types.QueryUnrelayedPacketsRequest{
+					PortID:    "test-port-id",
+					ChannelID: "",
+				}
+			},
+			false,
+		},
+		{
+			"invalid seq",
+			func() {
+				req = &types.QueryUnrelayedPacketsRequest{
+					PortID:    "test-port-id",
+					ChannelID: "test-channel-id",
+					Sequences: []uint64{0},
+				}
+			},
+			false,
+		},
+		{
+			"success",
+			func() {
+				_, _, _, _, channelA, _ := suite.coordinator.Setup(suite.chainA, suite.chainB)
+				ack := types.NewPacketAckCommitment(channelA.PortID, channelA.ID, 1, []byte("hash"))
+				suite.chainA.App.IBCKeeper.ChannelKeeper.SetPacketAcknowledgement(suite.chainA.GetContext(), channelA.PortID, channelA.ID, 1, ack.Hash)
```

Comment:
So I think UnrelayedPackets might not be working as intended. Basically we want a query function that can indicate if given a list of packet commitment sequences:
* does the packet commitment need to be relayed? (no ack exists)
* does the ack need to be relayed? (ack exists and packet commitment exists)

looking at this test, an ack is set, but no packet commitment is created for the corresponding ack (ack has been relayed, packet cycle complete), so the sequences returned should be empty.  

It's probably best to split the unrelayed packets into two funcs to avoid confusion or just add a bool in the request.

Reply:
> It's probably best to split the unrelayed packets into two funcs to avoid confusion or just add a bool in the request.

Do you want to split the command as well?

Reply:
yea that sounds good, a bool flag could work as well. Whichever is easier

Reply:
bool flag might be easier:
`gaia q ibc channel unrelayed-packets --acknowledgements true` would return all unrelayed acknowledgements